particle-builtins/src/builtins.rs

```
@@ -738,6 +753,51 @@ where
             }))
         }
     }
+
+    fn sign(&self, args: Args) -> Result<JValue, JError> {
+        let tetraplets = args.tetraplets;
+        let mut args = args.function_args.into_iter();
+        let data: Vec<u8> = Args::next("data", &mut args)?;
+
+        let error = "data should be passed from local registry.get_record_bytes()";
+        let tetraplet = tetraplets
+            .get(0)
+            .ok_or_else(|| JError::new(error.to_string()))?
+            .get(0)
+            .ok_or_else(|| JError::new(error.to_string()))?;
+
+        if tetraplet.peer_pk != self.local_peer_id.to_base58()
+            || tetraplet.service_id != "registry"
+            || tetraplet.function_name != "get_record_bytes"
+        {
+            return Err(JError::new(error.to_string()));
+        }
+
+        Ok(JValue::Array(
+            self.root_keypair
+                .sign(&data)?
+                .to_vec()
+                .into_iter()
+                .map(|n| json!(n))
+                .collect(),
+        ))
+    }
+
+    fn verify(&self, args: Args) -> Result<JValue, JError> {
+        let mut args = args.function_args.into_iter();
+        let signature: Vec<u8> = Args::next("signature", &mut args)?;
+        let data: Vec<u8> = Args::next("data", &mut args)?;
+        let signature =
+            Signature::from_bytes(self.root_keypair.public().get_key_format(), signature);
+
+        Ok(JValue::Bool(
+            self.root_keypair.public().verify(&data, &signature).is_ok(),
```

Maybe return an error message together with `false`?

---

We don't have it in JS and there is no meaningful message for verification failure

---

There are two cases where verification fails:
1. Signature bytes are of incorrect format
2. Signing equation is incorrect (ie signature forged or made for another data)

It's rather hard to tell one from the other when debugging cryptography code. One mistake in signature storage, and you always get this pesky `false` without ever knowing that the signature format is incorrect.

I don't know why, but even though you use `Signature::from_bytes`, according ed25519-dalek's code it's still possible to get a signature format error. See:
```rust
impl TryFrom<&ed25519::Signature> for InternalSignature {
    type Error = SignatureError;

    fn try_from(sig: &ed25519::Signature) -> Result<InternalSignature, SignatureError> {
        InternalSignature::from_bytes(sig.as_bytes())
    }
}
```

So I insist on having an error message along the result. Let's not allow previous decisions on API to be a handicap forever. At least without a very good reason.

---

Okay. 

@coder11 I will file an issue for Sig.verify return type change

---

DXJ-107